public/js/components/SourceTabs.js

```diff
@@ -30,45 +32,158 @@ function getFilename(url) {
-      className: classnames("source-tab", { active }),
-      key: source.get("id"),
-      onClick: () => selectSource(source.get("id"))
-    },
-    dom.div({ className: "filename" }, filename),
-    dom.div({ onClick: onClickClose },
-      Isvg({
-        className: "close-btn",
-        src: "images/close.svg",
-      })
-    )
-  );
+  const tabTopOffset = getTopOffset();
+  return sourceTabs.filter((tab, index) => {
+    return sourceTabEls[index].getBoundingClientRect().top > tabTopOffset;
+  });
 }
 
-function SourceTabs({ selectedSource, sourceTabs, selectSource, closeTab }) {
-  function renderTab(source) {
-    return sourceTab(source, selectedSource, selectSource, closeTab);
-  }
+const SourceTabs = React.createClass({
+  propTypes: {
+    sourceTabs: ImPropTypes.list,
+    selectedSource: ImPropTypes.map,
+    selectSource: PropTypes.func.isRequired,
+    closeTab: PropTypes.func.isRequired
+  },
 
-  return (
-    dom.div({ className: "source-tabs" },
-      isEnabled("features.tabs")
-        ? sourceTabs.map(renderTab)
-        : renderTab(selectedSource)
-    )
-  );
-}
+  displayName: "SourceTabs",
+
+  getInitialState() {
+    return {
+      dropdownShown: false,
+      hiddenSourceTabs: null
+    };
+  },
+
+  componentDidUpdate() {
+    this.updateHiddenSourceTabs(this.props.sourceTabs);
+  },
+
+  /*
+   * Updates the hiddenSourceTabs state, by
+   * finding the source tabs who have wrapped and are not on the top row.
+   */
+  updateHiddenSourceTabs(sourceTabs) {
+    const sourceTabEls = this.refs.sourceTabs.children;
+    const hiddenSourceTabs = getHiddenTabs(sourceTabs, sourceTabEls);
+
+    if (!hiddenSourceTabs.equals(this.state.hiddenSourceTabs)) {
+      this.setState({ hiddenSourceTabs });
+    }
+  },
+
+  toggleSourcesDropdown(e) {
+    this.setState({
+      dropdownShown: !this.state.dropdownShown,
+    });
+  },
+
+  renderSourcesDropdown() {
+    if (!this.state.hiddenSourceTabs) {
+      return dom.div({});
+    }
+
+    return dom.div({
+      className: "sources-dropdown dropdown",
+      ref: "sourcesDropdown",
+      style: { display: (this.state.dropdownShown ? "block" : "none") }
+    },
+      dom.ul({}, this.state.hiddenSourceTabs.map(this.renderDropdownSource))
+    );
+  },
+
+  renderDropdownSource(source) {
+    const { selectSource } = this.props;
+    const url = source && source.get("url");
+    const filename = getFilename(url);
+
+    return dom.li({
+      key: source.get("id"),
+      onClick: () => {
+        selectSource(source.get("id"), { position: 0 });
+        this.toggleSourcesDropdown();
+      }
+    }, filename);
+  },
+
+  renderSourcesDropdownButon() {
+    const hiddenSourceTabs = this.state.hiddenSourceTabs;
+    if (!hiddenSourceTabs || hiddenSourceTabs.size == 0) {
+      return dom.div({});
+    }
+
+    return dom.span(
+      {
+        className: "subsettings",
+        onClick: this.toggleSourcesDropdown
+      },
+      dom.img({ src: "images/subSettings.svg" })
```

Comment:
We might want a different icon. It's a little confusing to see 2 of those pretty close to each other. I think that icon is for settings specifically. If possible, use a placeholder for now.